File: commons/org.openmole.commons.aspect.eventdispatcher/src/main/java/org/openmole/commons/aspect/eventdispatcher/internal/EventDispatcher.java
/*
 * To change this template, choose Tools | Templates
 * and open the template in the editor.
 */
package org.openmole.commons.aspect.eventdispatcher.internal;

import java.util.Collections;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.logging.Level;
import java.util.logging.Logger;
import org.openmole.commons.aspect.eventdispatcher.IEventDispatcher;
import org.openmole.commons.aspect.eventdispatcher.IObjectChangedAsynchronousListener;
import org.openmole.commons.aspect.eventdispatcher.IObjectChangedAsynchronousListenerWithArgs;
import org.openmole.commons.aspect.eventdispatcher.IObjectChangedSynchronousListenerWithArgs;
import org.openmole.commons.aspect.eventdispatcher.IObjectChangedSynchronousListener;
import org.openmole.commons.aspect.eventdispatcher.IObjectConstructedAsynchronousListener;
import org.openmole.commons.aspect.eventdispatcher.IObjectConstructedSynchronousListener;
import org.openmole.commons.tools.service.Priority;
import org.openmole.commons.exception.InternalProcessingError;
import org.openmole.commons.exception.UserBadDataError;


/**
 *
 * @author reuillon
 */
public class EventDispatcher implements IEventDispatcher {

    final static Logger LOGGER = Logger.getLogger(EventDispatcher.class.getName());
    volatile int nb = 0;
    
    final static private ExecutorService Executor = Executors.newCachedThreadPool();

    final ObjectChangedListnerMap<IObjectChangedAsynchronousListener> asynchronousObjectChangedMap = new ObjectChangedListnerMap<IObjectChangedAsynchronousListener>();
    final ObjectChangedListnerMap<IObjectChangedSynchronousListener> synchronousObjectChangedMap = new ObjectChangedListnerMap<IObjectChangedSynchronousListener>();
    final ObjectChangedListnerMap<IObjectChangedAsynchronousListenerWithArgs> asynchronousObjectChangedWithArgsMap = new ObjectChangedListnerMap<IObjectChangedAsynchronousListenerWithArgs>();
    final ObjectChangedListnerMap<IObjectChangedSynchronousListenerWithArgs> synchronousObjectChangedWithArgsMap = new ObjectChangedListnerMap<IObjectChangedSynchronousListenerWithArgs>();
    final ObjectConstructedListnerMap<IObjectConstructedAsynchronousListener> asynchronousObjectConstructedMap = new ObjectConstructedListnerMap<IObjectConstructedAsynchronousListener>();
    final ObjectConstructedListnerMap<IObjectConstructedSynchronousListener> synchronousObjectConstructedMap = new ObjectConstructedListnerMap<IObjectConstructedSynchronousListener>();


    @Override
    public <T> void registerListener(T object, IObjectChangedAsynchronousListener<? super T> listner, String type) {
        asynchronousObjectChangedMap.registerListner(object, Priority.NORMAL.getValue(), listner, type);
    }

    @Override
    public void objectChanged(final Object object, final String type, final Object[] args) throws InternalProcessingError, UserBadDataError {
        
        final Iterable<IObjectChangedAsynchronousListener> objectChangedWithTypeAsynchronouslisteners = asynchronousObjectChangedMap.getListners(object, type);

        final Iterable<IObjectChangedAsynchronousListenerWithArgs> objectChangedWithTypeAsynchronouslistenersWithArgs = asynchronousObjectChangedWithArgsMap.getListners(object, type);


        //Avoid creating threads if no listners
        if (objectChangedWithTypeAsynchronouslisteners.iterator().hasNext()) {

            Executor.submit(new Runnable() {

                @Override
                public void run() {

                    /* --- Listners without args ---*/

                    synchronized (objectChangedWithTypeAsynchronouslisteners) {
                        for (IObjectChangedAsynchronousListener listner : objectChangedWithTypeAsynchronouslisteners) {
                            listner.objectChanged(object);
                        }
                    }

                    /* --- Listners with args ---*/

                    synchronized (objectChangedWithTypeAsynchronouslistenersWithArgs) {
                        for (IObjectChangedAsynchronousListenerWithArgs listner : objectChangedWithTypeAsynchronouslistenersWithArgs) {
                            listner.objectChanged(object, args);
                        }
                    }
                }
            });
        }
    
        /* --- Listners without args ---*/

        Iterable<IObjectChangedSynchronousListener> listeners = synchronousObjectChangedMap.getListners(object, type);

        synchronized (listeners) {
            for (IObjectChangedSynchronousListener listner : listeners) {
                listner.objectChanged(object);
            }
        }

        /* --- Listners with args ---*/

        Iterable<IObjectChangedSynchronousListenerWithArgs> listenersWithArgs = synchronousObjectChangedWithArgsMap.getListners(object, type);

        synchronized (listenersWithArgs) {
            for (IObjectChangedSynchronousListenerWithArgs listner : listenersWithArgs) {
                listner.objectChanged(object, args);
            }
        }

    }

    @Override
    public <T> void registerListener(T object, Integer priority, IObjectChangedSynchronousListener<? super T> listner, String type) {
        LOGGER.log(Level.FINE, "Register {0} {1}", new Object[]{nb++, object});
        synchronousObjectChangedMap.registerListner(object, priority, listner, type);
    }

    @Override
    public <T> void registerListener(Class<T> c, Integer priority, IObjectConstructedSynchronousListener<? super T> listner) {
        synchronousObjectConstructedMap.registerListner(c, priority, listner);
    }

    @Override
    public <T> void registerListener(Class<T> c, IObjectConstructedAsynchronousListener<? super T> listner) {
        asynchronousObjectConstructedMap.registerListner(c, Priority.NORMAL.getValue(), listner);
    }

    @Override
    public void objectConstructed(final Object object) {

        final Class c = object.getClass();
        final Iterable<IObjectConstructedAsynchronousListener> asynchronousObjectConstructedListners = asynchronousObjectConstructedMap.getListners(c);


        //Dont create thread if no assynchronous listner are registred
        if (asynchronousObjectConstructedListners.iterator().hasNext()) {
            Executor.submit(new Runnable() {

                @Override
                public void run() {
                    synchronized (asynchronousObjectConstructedListners) {
                        for (IObjectConstructedAsynchronousListener listner : asynchronousObjectConstructedListners) {
                            listner.objectConstructed(object);
                        }
                    }
                }
            });
        }

        Iterable<IObjectConstructedSynchronousListener> listeners = synchronousObjectConstructedMap.getListners(c);

        synchronized (listeners) {
            for (IObjectConstructedSynchronousListener listner : listeners) {
                listner.objectConstructed(object);
            }
        }
    }

    @Override
    public <T> void registerListener(T object, IObjectChangedAsynchronousListenerWithArgs<? super T> listner, String type) {
        LOGGER.log(Level.FINE, "Register {0} {1}", new Object[]{nb++, object});

        asynchronousObjectChangedWithArgsMap.registerListner(object, Priority.NORMAL.getValue(), listner, type);
    }

    @Override
    public <T> void registerListener(T object, Integer priority, IObjectChangedSynchronousListenerWithArgs<? super T> listner, String type) {
        LOGGER.log(Level.FINE, "Register {0} {1}", new Object[]{nb++, object});

        synchronousObjectChangedWithArgsMap.registerListner(object, priority, listner, type);
    }

    @Override
    public void objectChanged(Object object, String type) throws InternalProcessingError, UserBadDataError {
        LOGGER.log(Level.FINE, "Object changed {0}", object);
        objectChanged(object, type, Collections.EMPTY_LIST.toArray());
    }

}

File: commons/org.openmole.commons.aspect.eventdispatcher/src/main/scala/org/openmole/commons/aspect/eventdispatcher/internal/EventDispatcher.scala
/*
 * Copyright (C) 2010 reuillon
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package org.openmole.commons.aspect.eventdispatcher.internal

import java.util.concurrent.Executors

import org.openmole.commons.tools.service.Priority
import java.util.logging.Logger
import org.openmole.commons.aspect.eventdispatcher.IEventDispatcher
import org.openmole.commons.aspect.eventdispatcher.IObjectListener
import org.openmole.commons.aspect.eventdispatcher.IObjectListenerWithArgs

object EventDispatcher {
  
  //TODO maybe change to fixed thread pool some time 
  val Executor = Executors.newCachedThreadPool
  
}

class EventDispatcher extends IEventDispatcher {

  import EventDispatcher._
  
  val asynchronousObjectChangedMap = new ObjectListenerMap[IObjectListener[AnyRef]]
  val synchronousObjectChangedMap = new ObjectListenerMap[IObjectListener[AnyRef]]
  val asynchronousObjectChangedWithArgsMap = new ObjectListenerMap[IObjectListenerWithArgs[AnyRef]]
  val synchronousObjectChangedWithArgsMap = new ObjectListenerMap[IObjectListenerWithArgs[AnyRef]]
  val asynchronousObjectConstructedMap = new ClassListenerMap[IObjectListener[AnyRef]]
  val synchronousObjectConstructedMap = new ClassListenerMap[IObjectListener[AnyRef]]

  
  override def registerForObjectChangedAsynchronous[T](obj: T, listner: IObjectListener[T] , event: String) = {
    asynchronousObjectChangedMap.register(obj.asInstanceOf[AnyRef], Priority.NORMAL, listner.asInstanceOf[IObjectListener[AnyRef]], event)
  }
    
  override def registerForObjectChangedSynchronous[T](obj: T, priority: Int, listner: IObjectListener[T], event: String) = {  
    synchronousObjectChangedMap.register(obj.asInstanceOf[AnyRef], priority, listner.asInstanceOf[IObjectListener[AnyRef]], event)
    //Logger.getLogger(classOf[EventDispatcher].getName).fine("All listners " + synchronousObjectChangedMap.toString)  
  }
  
  
  override def registerForObjectChangedAsynchronous[T](obj: T, listner: IObjectListenerWithArgs[T] , event: String) = {
    asynchronousObjectChangedWithArgsMap.register(obj.asInstanceOf[AnyRef], Priority.NORMAL, listner.asInstanceOf[IObjectListenerWithArgs[AnyRef]], event)
  }
  
  override def registerForObjectChangedSynchronous[T](obj: T, priority: Int, listner: IObjectListenerWithArgs[T], event: String) = {
    synchronousObjectChangedWithArgsMap.register(obj.asInstanceOf[AnyRef], priority, listner.asInstanceOf[IObjectListenerWithArgs[AnyRef]], event)
  }
    
  override def registerForObjectConstructedSynchronous[T](c: Class[T], priority: Int, listner: IObjectListener[T]) = {
    synchronousObjectConstructedMap.register(c, priority, listner.asInstanceOf[IObjectListener[AnyRef]])
  }  
  
  override def registerForObjectConstructedAsynchronous[T](c: Class[T], listner: IObjectListener[T]) = {
    asynchronousObjectConstructedMap.register(c, Priority.NORMAL, listner.asInstanceOf[IObjectListener[AnyRef]])
  }
  
  
  override def objectChanged[T](obj: T, event: String, args: Array[Object]) = {
    //Logger.getLogger(classOf[EventDispatcher].getName).fine("Signal event " + event + " signaled from " + obj.toString)
    val objectChangedWithTypeAsynchronouslisteners = asynchronousObjectChangedMap.get(obj.asInstanceOf[AnyRef], event)
    val objectChangedWithTypeAsynchronouslistenersWithArgs = asynchronousObjectChangedWithArgsMap.get(obj.asInstanceOf[AnyRef], event)

    //Avoid creating threads if no listners
    if (objectChangedWithTypeAsynchronouslisteners.iterator.hasNext) {

      Executor.submit(new Runnable {

          override def run = {

            /* --- Listners without args ---*/

            objectChangedWithTypeAsynchronouslisteners.synchronized  {
              for (listner <- objectChangedWithTypeAsynchronouslisteners) {
                listner.eventOccured(obj.asInstanceOf[AnyRef])
              }
            }

            /* --- Listners with args ---*/

            objectChangedWithTypeAsynchronouslistenersWithArgs.synchronized {
              for (listner <- objectChangedWithTypeAsynchronouslistenersWithArgs) {
                listner.eventOccured(obj.asInstanceOf[AnyRef], args)
              }
            }
          }
        })
    }
    
    /* --- Listners without args ---*/

    val listeners = synchronousObjectChangedMap.get(obj.asInstanceOf[AnyRef], event)
     
    listeners.synchronized  {
      for (listner <- listeners) {
        //Logger.getLogger(classOf[EventDispatcher].getName).fine("Event no arg " + event + " from " + obj.toString + " signaled to " + listner.toString)
        listner.eventOccured(obj.asInstanceOf[AnyRef])
      }
    }

    /* --- Listners with args ---*/

    val listenersWithArgs = synchronousObjectChangedWithArgsMap.get(obj.asInstanceOf[AnyRef], event)

    listenersWithArgs.synchronized {
      for (listner <- listenersWithArgs) {
        //Logger.getLogger(classOf[EventDispatcher].getName).fine("Event no arg " + event + " from " + obj.toString + " signaled to " + listner.toString)
        listner.eventOccured(obj.asInstanceOf[AnyRef], args)
      }
    }
    
  }
   
  override def objectChanged[T](obj: T, event: String) = objectChanged(obj.asInstanceOf[AnyRef], event, Array.empty)

  override def objectConstructed(obj: Object) = {

    val c = obj.getClass
    val asynchronousObjectConstructedListners = asynchronousObjectConstructedMap.get(c)


    //Dont create thread if no assynchronous listner are registred
    if (asynchronousObjectConstructedListners.iterator.hasNext) {
      Executor.submit(new Runnable() {

          override def run = {
            asynchronousObjectConstructedListners.synchronized  {
              for (listner <- asynchronousObjectConstructedListners) {
                listner.eventOccured(obj)
              }
            }
          }
        })
    }

    val listeners = synchronousObjectConstructedMap.get(c)

    listeners.synchronized {
      for (listner <- listeners) {
        listner.eventOccured(obj)
      }
    }
  }

}
